File: Demo/ObjcDemo/ComplexPageController.m
//
//  ComplexPageController.m
//  ObjcDemo
//
//  Created by kelin on 2020/11/10.
//

#import "ComplexPageController.h"
#import "KKAutoCycleScrollView.h"
#import <SDWebImage/SDWebImage.h>

@interface ComplexPageController ()<KKAutoCycleScrollViewDelegate, KKAutoCycleScrollViewDataSource>
@property (nonatomic, strong) KKAutoCycleScrollView *cycleView1;
@property (nonatomic, strong) KKAutoCycleScrollView *cycleView2;
@property (nonatomic, strong) UIPageControl *pageControl1;
@property (nonatomic, strong) UIPageControl *pageControl2;
@property (nonatomic, strong) UILabel *titleLabel;
@end

@implementation ComplexPageController

NSArray *cycle1Datas() {
    return @[
        @{@"image":@"1.png", @"title":@"这是第一张图片"},
        @{@"image":@"2.png", @"title":@"这是第二张图片"},
        @{@"image":@"3.png", @"title":@"这是第三张图片"},
        @{@"image":@"4.png", @"title":@"这是第四张图片"},
    ];
}

NSArray *cycle2Datas() {
    return @[
        @{@"image":@"https://icweiliimg6.pstatp.com/weili/l/378712400698146835.jpg", @"title":@"第一张网络图片【草原】"},
        @{@"image":@"https://icweiliimg6.pstatp.com/weili/l/903360033333313563.jpg", @"title":@"第二张网络图片【公路】"},
        @{@"image":@"https://weiliicimg1.pstatp.com/weili/l/812099039242354699.jpg", @"title":@"第三张网络图片【灯笼】"},
        @{@"image":@"https://icweiliimg1.pstatp.com/weili/l/812022923831934995.jpg", @"title":@"第四张网络图片【雪山】"},
    ];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor whiteColor];
    self.title = @"复杂轮播图";
    [self initCycleView1];
    [self initCycleView2];
    
    [self.cycleView1 scrollToIndex:1 animated:NO];
}

- (void)initCycleView1 {
    _cycleView1 = [[KKAutoCycleScrollView alloc] initWithFrame:CGRectMake(0, 100, self.view.frame.size.width, 200)];
    _cycleView1.delegate = self;
    _cycleView1.dataSource = self;
    _cycleView1.autoCycle = NO;
    
    [_cycleView1 create];
    
    _pageControl1 = [[UIPageControl alloc] initWithFrame:CGRectMake((_cycleView1.frame.size.width - 200)/2, CGRectGetMaxY(_cycleView1.frame)- 40, 200, 50)];
    _pageControl1.pageIndicatorTintColor = [UIColor whiteColor];
    _pageControl1.currentPageIndicatorTintColor = [UIColor blueColor];
    _pageControl1.numberOfPages = cycle1Datas().count;
    [_pageControl1 addTarget:self action:@selector(pageChange:) forControlEvents:UIControlEventValueChanged];
    if (@available(iOS 14.0, *)) {
        _pageControl1.allowsContinuousInteraction = YES;
    }
    
    [self.view addSubview:_cycleView1];
    [self.view addSubview:_pageControl1];
}

- (void)initCycleView2 {
    _cycleView2 = [[KKAutoCycleScrollView alloc] initWithFrame:CGRectMake(0, CGRectGetMaxY(_cycleView1.frame) + 30, self.view.frame.size.width, 200)];
    _cycleView2.delegate = self;
    _cycleView2.dataSource = self;
    _cycleView2.duration = 2;
    _cycleView2.autoCycle = YES;
    
    _titleLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, CGRectGetMaxY(_cycleView2.frame)- 40, _cycleView2.frame.size.width, 40)];
    _titleLabel.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.5];
    _titleLabel.textColor = [UIColor whiteColor];
    
    _pageControl2 = [[UIPageControl alloc] initWithFrame:CGRectMake(_cycleView2.frame.size.width - 150, CGRectGetMaxY(_cycleView2.frame)- 40, 200, 40)];
    _pageControl2.pageIndicatorTintColor = [UIColor whiteColor];
    _pageControl2.userInteractionEnabled = NO;
    _pageControl2.currentPageIndicatorTintColor = [UIColor blueColor];
    _pageControl2.numberOfPages = cycle2Datas().count;
    
    [self.view addSubview:_cycleView2];
    [self.view addSubview:_titleLabel];
    [self.view addSubview:_pageControl2];
    
    [_cycleView2 create];

}

- (void)pageChange:(UIPageControl *)pageControl {
    if ([self pageControlCanContinuousInteraction:pageControl]) {
        [self.cycleView1 scrollToIndex:pageControl.currentPage animated:NO];
    } else {
        [self.cycleView1 scrollToIndex:pageControl.currentPage animated:YES];
    }
}

- (BOOL)pageControlCanContinuousInteraction:(UIPageControl *)pageControl {
    if (@available(iOS 14.0, *)) {
        if (!pageControl.allowsContinuousInteraction) {
            return NO;
        }
        NSArray *gestures = [pageControl gestureRecognizers];
        for (UIGestureRecognizer *gesture in gestures) {
            if (gesture.state == UIGestureRecognizerStateFailed ||
                gesture.state == UIGestureRecognizerStateEnded) {
                return NO;
            }
        }
        return YES;
    }
    return NO;
}


- (void)cycleView:(KKAutoCycleScrollView *)cycleView addSubViewsInPageView:(UIView *)pageView subViewAtIndex:(NSInteger)index {
    
    if (cycleView == self.cycleView1) {
        UIImageView *imageView = [[UIImageView alloc] initWithFrame:pageView.bounds];
        imageView.image = [UIImage imageNamed:cycle1Datas()[index][@"image"]];
        [pageView addSubview:imageView];
        
        UILabel *descLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, pageView.frame.size.height - 40, pageView.frame.size.width, 40)];
        descLabel.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.5];
        descLabel.text = cycle1Datas()[index][@"title"];
        descLabel.textColor = [UIColor whiteColor];
        [pageView addSubview:descLabel];
    }
    
    
    if (cycleView == self.cycleView2) {
        UIImageView *imageView = [[UIImageView alloc] initWithFrame:pageView.bounds];
        [imageView sd_setImageWithURL:[NSURL URLWithString:cycle2Datas()[index][@"image"]]];
        [pageView addSubview:imageView];
    }
    

}

- (NSInteger)numberOfViewCountInCycleView:(KKAutoCycleScrollView *)cycleView {
    if (cycleView == self.cycleView1) {
        return cycle1Datas().count;
    }
    if (cycleView == self.cycleView2) {
        return cycle2Datas().count;
    }
    return 0;
}

- (void)cycleView:(KKAutoCycleScrollView *)cycleView didClickPageViewAtIndex:(NSInteger)index {
    NSLog(@"current page is %ld", index);
}

- (void)cycleView:(KKAutoCycleScrollView *)cycleView outerViewScrollingIndex:(NSInteger)index {
    if (cycleView == self.cycleView1) {
        self.pageControl1.currentPage = index;
    }
    
    if (cycleView == self.cycleView2) {
        self.titleLabel.text = cycle2Datas()[index][@"title"];
        self.pageControl2.currentPage = index;
    }
}

- (void)dealloc {
    [_cycleView1 destory];
    [_cycleView2 destory];
}



@end

File: Demo/SwiftDemo/ComplexPageController.swift
//
//  ComplexPageController.swift
//  SwiftDemo
//
//  Created by kelin on 2020/11/10.
//

import Foundation
import SDWebImage

class ComplexPageController: UIViewController, KKAutoCycleScrollViewDelegate, KKAutoCycleScrollViewDataSource {
    
    var cycleView1:KKAutoCycleScrollView!
    var cycleView2:KKAutoCycleScrollView!
    var pageContrl1:UIPageControl!
    var pageContrl2:UIPageControl!
    var titleLabel:UILabel!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.title = "网络图片加载"
        self.view.backgroundColor = UIColor.white
        
        self.initCycleView1()
        self.initCycleView2()
        self.cycleView1.scrollToIndex(index: 1, animated: false)
    }
    
    func initCycleView1() -> Void {

        self.cycleView1 = KKAutoCycleScrollView(frame: CGRect(x: 0, y: 100, width: self.view.frame.size.width, height: 200))
        self.cycleView1.delegate = self;
        self.cycleView1.dataSource = self;
        self.cycleView1.autoCycle = false
        
        self.pageContrl1 = UIPageControl(frame: CGRect(x: (self.cycleView1.frame.size.width - 200)/2, y: self.cycleView1.frame.maxY - 40, width: 200, height: 50))
        self.pageContrl1.pageIndicatorTintColor = UIColor.white
        self.pageContrl1.currentPageIndicatorTintColor = UIColor.blue
        self.pageContrl1.numberOfPages = TestData.cycle1Datas().count
        self.pageContrl1.addTarget(self, action: #selector(ComplexPageController.pageChange(_:)), for: .valueChanged)
        if #available(iOS 14.0, *) {
            self.pageContrl1.allowsContinuousInteraction = true
        }
        
        self.view.addSubview(self.cycleView1)
        self.view.addSubview(self.pageContrl1)
        
        self.cycleView1.create()
    }
    
    func initCycleView2() -> Void {

        self.cycleView2 = KKAutoCycleScrollView(frame: CGRect(x: 0, y: self.cycleView1.frame.maxY + 30, width: self.view.frame.size.width, height: 200))
        self.cycleView2.delegate = self;
        self.cycleView2.dataSource = self;
        self.cycleView2.autoCycle = true
        self.cycleView2.create()
        
        self.titleLabel = UILabel(frame: CGRect(x: 0, y: self.cycleView2.frame.maxY - 40, width: self.cycleView2.frame.size.width, height: 40))
        self.titleLabel.backgroundColor = UIColor.black.withAlphaComponent(0.5)
        self.titleLabel.textColor = UIColor.white
        
        self.pageContrl2 = UIPageControl(frame: CGRect(x: self.cycleView2.frame.size.width - 150, y: self.cycleView2.frame.maxY - 40, width: 200, height: 40))
        self.pageContrl2.pageIndicatorTintColor = UIColor.white
        self.pageContrl2.currentPageIndicatorTintColor = UIColor.blue
        self.pageContrl2.numberOfPages = TestData.cycle2Datas().count
        self.pageContrl2.isUserInteractionEnabled = false
        self.pageContrl2.addTarget(self, action: #selector(ComplexPageController.pageChange(_:)), for: .valueChanged)
        
        self.view.addSubview(self.cycleView2)
        self.view.addSubview(self.titleLabel)
        self.view.addSubview(self.pageContrl2)
        
        self.cycleView2.create()
    }
    
    @objc func pageChange(_ sender:UIPageControl) -> Void {
        if self.pageControlCanContinuousInteraction(pageControl: sender) {
            self.cycleView1.scrollToIndex(index: sender.currentPage, animated: false)
        } else {
            self.cycleView1.scrollToIndex(index: sender.currentPage, animated: true)
        }
    }
    
    private func pageControlCanContinuousInteraction(pageControl:UIPageControl) -> Bool {
        if #available(iOS 14.0, *) {
            if !pageControl.allowsContinuousInteraction {
                return false
            }
            let gestures:[UIGestureRecognizer] = pageControl.gestureRecognizers!
            for gesture in gestures {
                if gesture.state == .failed || gesture.state == .ended {
                    return false
                }
            }
            return true
        }
        return false
    }
    
    /**
     
     - (void)pageChange:(UIPageControl *)pageControl {
         if ([self pageControlCanContinuousInteraction:pageControl]) {
             [self.cycleView1 scrollToIndex:pageControl.currentPage animated:NO];
         } else {
             [self.cycleView1 scrollToIndex:pageControl.currentPage animated:YES];
         }
     }

     - (BOOL)pageControlCanContinuousInteraction:(UIPageControl *)pageControl {
         if (@available(iOS 14.0, *)) {
             if (!pageControl.allowsContinuousInteraction) {
                 return NO;
             }
             NSArray *gestures = [pageControl gestureRecognizers];
             for (UIGestureRecognizer *gesture in gestures) {
                 if (gesture.state == UIGestureRecognizerStateFailed ||
                     gesture.state == UIGestureRecognizerStateEnded) {
                     return NO;
                 }
             }
             return YES;
         }
         return NO;
     }
     
     */
    
    func numberOfViewCountInCycleView(cycle: KKAutoCycleScrollView) -> NSInteger {
        if cycle == self.cycleView1 {
            return TestData.cycle1Datas().count
        }
        if cycle == self.cycleView2 {
            return TestData.cycle2Datas().count
        }
        return 0
    }
    
    func cycleViewAddSubViewsInPageViewAtIndex(cycle: KKAutoCycleScrollView, pageView: UIView, index: NSInteger) {
        if cycle == self.cycleView1 {
            let imageView = UIImageView(frame: pageView.bounds)
            imageView.image = UIImage(named: TestData.cycle1Datas()[index]["image"]!)
            pageView.addSubview(imageView)
            
            let label = UILabel(frame: CGRect(x: 0, y: pageView.frame.size.height - 40, width: pageView.frame.size.width, height: 40))
            label.backgroundColor = UIColor.black.withAlphaComponent(0.5)
            label.text = TestData.cycle1Datas()[index]["title"]
            label.textColor = UIColor.white
            pageView.addSubview(label)
        }
        
        if cycle == self.cycleView2 {
            let imageView = UIImageView(frame: pageView.bounds)
            imageView.sd_setImage(with: URL(string: TestData.cycle2Datas()[index]["image"]!), completed: nil)
            pageView.addSubview(imageView)
        }
    }
    
    func cycleViewDidClickPageViewAtIndex(cycleView: KKAutoCycleScrollView, index: NSInteger) {
        print("current page is %ld", index)
    }
    
    func cycleViewOuterViewScrollingIndex(cycleView: KKAutoCycleScrollView, index: NSInteger) {
        if cycleView == self.cycleView1 {
            self.pageContrl1.currentPage = index
        }
        
        if cycleView == self.cycleView2 {
            guard self.titleLabel != nil else {return}
            self.titleLabel.text = TestData.cycle2Datas()[index]["title"]!
            self.pageContrl2.currentPage = index
        }
    }
    
    deinit {
        self.cycleView1.destory()
        self.cycleView2.destory()
    }
}
